src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -4298,5 +4298,598 @@ class B
+      IL_001c:  stloc.0
+      IL_001d:  ldloc.0
+      IL_001e:  stloc.1
+      IL_001f:  ldloc.1
+      IL_0020:  brfalse.s  IL_0026
+      IL_0022:  ldc.i4.1
+      IL_0023:  stloc.2
+      IL_0024:  br.s       IL_002a
+      IL_0026:  ldc.i4.0
+      IL_0027:  stloc.2
+      IL_0028:  br.s       IL_002a
+      IL_002a:  ldloc.2
+      IL_002b:  ret
+    }
+");
+            compVerifier.VerifyIL("C.M2", @"
+    {
+      // Code size       44 (0x2c)
+      .maxstack  2
+      .locals init (bool V_0,
+                    int V_1)
+      IL_0000:  nop
+      IL_0001:  ldarg.0
+      IL_0002:  ldc.i4.s   65
+      IL_0004:  blt.s      IL_000b
+      IL_0006:  ldarg.0
+      IL_0007:  ldc.i4.s   90
+      IL_0009:  ble.s      IL_001d
+      IL_000b:  ldarg.0
+      IL_000c:  ldc.i4.s   97
+      IL_000e:  blt.s      IL_001a
+      IL_0010:  ldarg.0
+      IL_0011:  ldc.i4.s   122
+      IL_0013:  cgt
+      IL_0015:  ldc.i4.0
+      IL_0016:  ceq
+      IL_0018:  br.s       IL_001b
+      IL_001a:  ldc.i4.0
+      IL_001b:  br.s       IL_001e
+      IL_001d:  ldc.i4.1
+      IL_001e:  stloc.0
+      IL_001f:  ldloc.0
+      IL_0020:  brfalse.s  IL_0026
+      IL_0022:  ldc.i4.1
+      IL_0023:  stloc.1
+      IL_0024:  br.s       IL_002a
+      IL_0026:  ldc.i4.0
+      IL_0027:  stloc.1
+      IL_0028:  br.s       IL_002a
+      IL_002a:  ldloc.1
+      IL_002b:  ret
+    }
+");
+
+            compilation = CreateCompilation(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.RegularWithPatternCombinators);
+            compilation.VerifyDiagnostics();
+            compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+            compVerifier.VerifyIL("C.M1", @"
+    {
+      // Code size       33 (0x21)
+      .maxstack  2
+      .locals init (bool V_0)
+      IL_0000:  ldarg.0
+      IL_0001:  ldc.i4.s   65
+      IL_0003:  blt.s      IL_0018
+      IL_0005:  ldarg.0
+      IL_0006:  ldc.i4.s   90
+      IL_0008:  ble.s      IL_0014
+      IL_000a:  ldarg.0
+      IL_000b:  ldc.i4.s   97
+      IL_000d:  blt.s      IL_0018
+      IL_000f:  ldarg.0
+      IL_0010:  ldc.i4.s   122
+      IL_0012:  bgt.s      IL_0018
+      IL_0014:  ldc.i4.1
+      IL_0015:  stloc.0
+      IL_0016:  br.s       IL_001a
+      IL_0018:  ldc.i4.0
+      IL_0019:  stloc.0
+      IL_001a:  ldloc.0
+      IL_001b:  brfalse.s  IL_001f
+      IL_001d:  ldc.i4.1
+      IL_001e:  ret
+      IL_001f:  ldc.i4.0
+      IL_0020:  ret
+    }
+");
+            compVerifier.VerifyIL("C.M2", @"
+    {
+      // Code size       24 (0x18)
+      .maxstack  2
+      IL_0000:  ldarg.0
+      IL_0001:  ldc.i4.s   65
+      IL_0003:  blt.s      IL_000a
+      IL_0005:  ldarg.0
+      IL_0006:  ldc.i4.s   90
+      IL_0008:  ble.s      IL_0014
+      IL_000a:  ldarg.0
+      IL_000b:  ldc.i4.s   97
+      IL_000d:  blt.s      IL_0016
+      IL_000f:  ldarg.0
+      IL_0010:  ldc.i4.s   122
+      IL_0012:  bgt.s      IL_0016
```

Comment:
Seems like the codegen from this point on is simpler in M2 than M1. Is this something that matters in practice?

Reply:
It is unavoidable.  The compiler has the freedom to reorder operations in M1 and skip operations it can infer do not affect the result, but not in M2 as the order of operations is specified.

---
In reply to: [388690565](https://github.com/dotnet/roslyn/pull/42109#discussion_r388690565) [](ancestors = 388690565)